text: allow customtype to have a Bytes() method

tamird opened the pull request:
If encoding.TextMarshaler is implemented, it will be preferred.

Related to #199. @awalterschulze cc @bdarnell

awalterschulze commented:
I think the fact that time.Time also implements encoding.TextMarshaler is causing the tests to fail.

Also how about a test which includes a customtype with a Bytes method.

You can also add yourself or your company to the AUTHORS file if you want some copyright.

tamird commented:
ugh, that's pretty annoying. any ideas on how to fix this?

awalterschulze commented:
How about simply checking whether the type is time.Time or time.Duration and reverting to the default behaviour for those?

You can use props.StdTime and props.StdDuration to check.

tamird commented:
@awalterschulze implemented your suggestion, though I don't feel good about it :(

awalterschulze commented:
Do you have a better plan?
It doesn't look that bad to me.

tamird commented:
Nah, this is probably fine. Merge at will!

tamird commented:
Er, hang on, I think we still need a test.

awalterschulze commented:
Yes please :)

tamird commented:
Wrote a test! Unfortunately, `raw` expects a serialized proto. Opened https://github.com/golang/protobuf/issues/311 to discuss removal, but I think we can just go ahead and remove it from gogo/protobuf. WDYT?

tamird commented:
Pushed an update; let me know if that looks good to you and I'll squash it down.

awalterschulze commented:
The build is failing because you haven't merged in a long time.
Go1.6 failures are permitted.

awalterschulze commented:
Maybe I am missing something, but this diff is look really weird.
It seems like you have only deleted some code concerning raw and added a bytes method to some Uuid types.  Maybe my memory is wonky, but I remember this pull request going a totally different way with extra checking for stdtime and stdduration.

awalterschulze commented:
Lets first see what happens with golang/protobuf before simply removing it.

tamird commented:
> It seems like you have only deleted some code concerning raw and added a bytes method to some Uuid types. Maybe my memory is wonky, but I remember this pull request going a totally different way with extra checking for stdtime and stdduration.

Right; the old approach was to whitelist `stdtime` and `stdduration` so that they don't fall into the textmarshaler case, but that's no longer necessary now that `raw` is simply removed. I added `Bytes() []byte` to `Uuid` to demonstrate that the `Bytes() []byte` method is no longer treated specially.

awalterschulze commented:
Yes but then I loose a feature that golang/protobuf has.
I have not done this before.
Not without the user explicitly asking for it using an extension of vanity binary.

tamird commented:
What's the feature? The ability to pass "raw" messages? It's not documented, and according to @bcmills exists only to support an internal Google use case.

awalterschulze commented:
Being used by google internally is not not a goal of gogoprotobuf.

tamird commented:
I know! That's what I'm saying; `raw` is only used by Google internally.

awalterschulze commented:
DId you spot my double negative?

tamird commented:
Heh, I guess I did not. Well, I guess that's awkward, then.

tamird commented:
Note that https://github.com/golang/protobuf/issues/311 was "fixed" in https://github.com/golang/protobuf/pull/482, which will be merged to master eventually (https://github.com/golang/protobuf/issues/479). Once that happens, and the changes are pulled into gogo/protobuf, I'll rebase this.

awalterschulze commented:
I will also create a dev branch, where I will start merging those changes into.  Hopefully that will also help.

tamird commented:
Ok. Let me know when that branch is ready and I'll fix this up.

On Feb 2, 2018 11:53, "Walter Schulze" <notifications@github.com> wrote:

I will also create a dev branch, where I will start merging those changes
into. Hopefully that will also help.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/gogo/protobuf/pull/227#issuecomment-362636372>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABdsPD_8HyGJc09C_NTyhh5kQ_qJgnimks5tQzligaJpZM4KwPnF>
.

awalterschulze commented:
The dev branch is finally ready :)

tamird commented:
OK, rebased. I think there's really nothing to do here, now, since upstream removed their overly broad type. All that's left here is a test.

awalterschulze commented:
Is the writeRaw function now unused code? Maybe the removal of this
function should be upstreamed?

On Sun, 8 Apr 2018, 23:38 Tamir Duberstein, <notifications@github.com>
wrote:

> OK, rebased. I think there's really nothing to do here, now, since
> upstream removed their overly broad type. All that's left here is a test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gogo/protobuf/pull/227#issuecomment-379584727>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABvsLdbmSv0WXnpgFqmLmRc1-ctzZMTcks5tmoNBgaJpZM4KwPnF>
> .
>

awalterschulze commented:
I think there are already tests for Uuid and timestamp that do text marshaling:

  - TestNidOptCustomProtoText
  - TestStdTypesProtoText

So by you adding a Bytes methods to Uuid, I think it is getting tested.

awalterschulze commented:
So if you can upstream writeRaw instead of removing it here and add a comment to the Bytes method for why it is there, just so that we don't remove it, thinking it is something trivial.  Then this is ready to merge, no test needed :)

awalterschulze commented:
I have pulled it in :)

tamird commented:
Updated.

awalterschulze commented:
Can't believe this is finally fixed.
Thank you for your perseverance
